---
title: "ABC inferences"
output: 
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{ABC Inferences with MethEvolSIM}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r, echo=FALSE}
## TODO: Update 
library(devtools)
load_all()
```


# Worflow for simulation-based inferences

This workflow uses [Snakemake](https://snakemake.readthedocs.io) to manage the steps for simulating DNA methylation patterns, assessing ABC accuracy, and inferring model parameters.

## Requirements

- Snakemake installed (`conda create -c conda-forge -c bioconda -n snakemake snakemake`)

- R packages: optparse (add to dependencies), parallel, ape

- Access to the provided `Snakefile` and associated scripts.

- An .RData file with a dataframe specifying the distribution of methylation sites in the genomic region.

- An .RData file with a tree in ape format (TODO: modify this and check whether scrm is also producing trees in the same format)

## Steps in the Workflow

1. Save as simulation design the genomic structure, tree(s) -- GIVEN -- and sampled parameter combinations.

2. Run the simulations. TODO: Before submitting load package without devtools

3. Compute the selected summary statistics. TODO: When added sumStats functions to package, stop sourcing script.

4. Assess ABC accuracy using cross-validation.

5. Infer parameters from empirical DNA methylation data.

## Selection of the genomic region

1. [Download](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000238955.4/) the *M. zebra* reference genome (GCF_000238955.4_M_zebra_UMD2a_genomic.fa) from the NCBI Assembly database. This genome is part of the M. zebra UMD2a assembly, which includes associated data such as gene annotations and other resources.

2. Transform CGI annotation (add citation) in txt format to BED

```{bash, eval = FALSE}
awk 'NR > 1 {print $1 "\t" $2-1 "\t" $3}' CGI-Mzebra.txt > CGI-Mzebra.bed
```

