---
title: "Summary Statistics"
output: 
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{Summary Statistics with MethEvolSIM}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r, echo=FALSE}
## TODO: Update 
library(devtools)
load_all()
```


Interpretation of arguments: 

 - tip can be interpreted as sample or replicate
 
 - structure can be interpreted as island or non-island

## Frequency of methylation states per genomic structure type

### Global Mean

- **Partially-methylated sites**: Let $x_{i,t}$ and $y_{j,t}$ be the frequency of methylation state $p$ in island $i \in\{1, \cdots, I\}$ and non-island $j \in\{1, \cdots, J\}$ at tip $t$, respectively. We define the mean global frequency of $x$ and $y$ as:

\[
\overline{x}=\frac1I\sum_{i=1}^{I}\frac1T\sum_{t=1}^{T}x_{i,t} \\
\overline{y}=\frac1J\sum_{j=1}^{J}\frac1T\sum_{t=1}^{T}y_{j,t}
\]

, where $x_{i,t}$ is the frequency of state $p$ in island $i$ and tip $t$ and $y_{j,t}$ is the frequency of state $p$ in non-island $j$ and tip $t$. 

- **Methylated sites**: Let $w_{i,t}$ and $z_{j,t}$ be the frequency of methylation state $m$ in island $i \in\{1, \cdots, I\}$ and non-island $j \in\{1, \cdots, J\}$ at tip $t$, respectively. We define the mean global frequency of $w$ and $z$ as:

\[
\overline{w}=\frac1I\sum_{i=1}^{I}\frac1T\sum_{t=1}^{T}w_{i,t} \\
\overline{z}=\frac1J\sum_{j=1}^{J}\frac1T\sum_{t=1}^{T}z_{j,t}
\]

, where $w_{i,t}$ is the frequency of state $p$ in island $i$ and tip $t$ and $z_{j,t}$ is the frequency of state $m$ in non-island $j$ and tip $t$. 


These values can be computed with the functions `get_islandMeanFreqP`, `get_nonislandMeanFreqP`,`get_islandMeanFreqM` and `get_nonislandMeanFreqM` as in the following examples:

```{r}
# 1 tip / sample / replicate
sample_n <- 1
index_islands <- c(1, 3)
index_nonislands <- c(2, 4)
data <- list(c(.5, .5, 0, 0, 0, .5), c(.5, 0, 0, .5), c(.5, .5, 0), c(0, 0, .5)) # tip 1
get_islandMeanFreqP(index_islands, data, sample_n)
get_nonislandMeanFreqP(index_nonislands, data, sample_n)
get_islandMeanFreqM(index_islands, data, sample_n)
get_nonislandMeanFreqM(index_nonislands, data, sample_n)
```

```{r}
# 2 tip / sample / replicate
sample_n <- 2
index_islands <- c(1, 3)
index_nonislands <- c(2, 4)
data <- list(
  list(c(.5, .5, 0, 0, 0, .5), c(.5, 0, 0, .5), c(.5, .5, 0), c(0, 0, .5)), # tip 1
  list(c(0, .5, .5, 1, 1, .5), c(1, .5, 1, .5), c(0, .5, .5), c(1, .5, 1)) # tip 2
  )
get_islandMeanFreqP(index_islands, data, sample_n)
get_nonislandMeanFreqP(index_nonislands, data, sample_n)
get_islandMeanFreqM(index_islands, data, sample_n)
get_nonislandMeanFreqM(index_nonislands, data, sample_n)
```

### Mean SD of observed frequencies per structure type at tree tips

- **Partially-methylated sites**: Let $x_{i,t}$ and $y_{j,t}$ be the frequency of methylation state $p$ in island $i \in\{1, \cdots, I\}$ and non-island $j \in\{1, \cdots, J\}$, respectively. We define the standard deviation of the frequency of $x$ and $y$ at tip $t$ as:

$$
\sigma_{t}(x)=\sqrt{\frac{1}{I-1}\sum_{i=1}^{I}(x_{i,t}-\overline{x_{.,t}})^2} \\
\sigma_{t}(y)=\sqrt{\frac{1}{J-1}\sum_{j=1}^{J}(y_{i,t}-\overline{y_{.,t}})^2}
$$


, where $\overline{x_{.,t}}$ is the mean frequency of state $p$ at tip $t$ in islands and $\overline{y_{.,t}}$ is the mean frequency of state $p$ at tip $t$ in non-islands:

$$
\overline{x_{.,t}}=\frac{1}{I}\sum_{i=1}^{I}x_{i,t} \\
\overline{y_{.,t}}=\frac{1}{J}\sum_{j=1}^{J}x_{j,t}
$$ 

Therefore, the mean standard deviation across tips is:

$$
\hat{\sigma}(x)=\frac{1}{T}\sum_{t=1}^{T}\sigma_t(x) \\
\hat{\sigma}(y)=\frac{1}{T}\sum_{t=1}^{T}\sigma_t(y) \\
$$


- **Methylated sites**: Let $w_{i,t}$ and $z_{j,t}$ be the frequency of methylation state $m$ in island $i \in\{1, \cdots, I\}$ and non-island $j \in\{1, \cdots, J\}$, respectively. We define the standard deviation of the frequency of $w$ and $z$ at tip $t$ as:

$$
\sigma_{t}(w)=\sqrt{\frac{1}{I-1}\sum_{i=1}^{I}(x_{i,t}-\overline{x_{.,t}})^2} \\
\sigma_{t}(z)=\sqrt{\frac{1}{J-1}\sum_{j=1}^{J}(y_{i,t}-\overline{y_{.,t}})^2}
$$


, where $\overline{x_{.,t}}$ is the mean frequency of state $m$ at tip $t$ in islands and $\overline{z_{.,t}}$ is the mean frequency of state $m$ at tip $t$ in non-islands:

$$
\overline{w_{.,t}}=\frac{1}{I}\sum_{i=1}^{I}w_{i,t} \\
\overline{z_{.,t}}=\frac{1}{J}\sum_{j=1}^{J}z_{j,t}
$$

Therefore, the mean standard deviation across tips is:

$$
\hat{\sigma}(w)=\frac{1}{T}\sum_{t=1}^{T}\sigma_t(w) \\
\hat{\sigma}(z)=\frac{1}{T}\sum_{t=1}^{T}\sigma_t(z) \\
$$
These values can be computed with the functions `get_islandSDFreqP`, `get_nonislandSDFreqP`,`get_islandSDFreqM` and `get_nonislandSDFreqM` as in the following examples:

```{r}
# 1 tip / sample / replicate
sample_n <- 1
index_islands <- c(1, 3)
index_nonislands <- c(2, 4)
data <- list(c(.5, .5, 0, 1, 1, .5), c(.5, 0, 1, .5), c(.5, .5, 0), c(0, 0, .5)) # tip 1
get_islandSDFreqP(index_islands, data, sample_n)
get_nonislandSDFreqP(index_nonislands, data, sample_n)
get_islandSDFreqM(index_islands, data, sample_n)
get_nonislandSDFreqM(index_nonislands, data, sample_n)
```

```{r}
# 2 tip / sample / replicate
sample_n <- 2
index_islands <- c(1, 3)
index_nonislands <- c(2, 4)
data <- list(
  list(c(.5, .5, 0, 0, 0, 1), c(.5, 0, 0, .5), c(1, .5, 0), c(0, 0, .5)), # tip 1
  list(c(0, .5, .5, 1, 1, .5), c(1, .5, 1, .5), c(0, .5, .5), c(1, .5, 1)) # tip 2
  )
get_islandSDFreqP(index_islands, data, sample_n)
get_nonislandSDFreqP(index_nonislands, data, sample_n)
get_islandSDFreqM(index_islands, data, sample_n)
get_nonislandSDFreqM(index_nonislands, data, sample_n)
```

Note that we average across tips as there can be correlations between the mean standard deviation of tips (e.g. under the case of an event affecting the states at two tips).

# Mean correlation of a central segment of methylation states per genomic structure type

Let $s$ be the length of the 2 adjacent segments of central positions to consider.
Let $l$ represent a number of positions to exclude at the start and end of each island/non-island structure.
Define the index positions $p=l+1$ and $p' = s-l$.
Let $x_{i,t,[p,p'-1]}$ and $x_{i,t,[p+1,p']}$ be the methylation states $x \in \{u, p, m\}$ in island $i \in \{1, \cdots ,I\}$ at tip $t$ for the given segments. 
Let $y_{j,t,[p,p'-1]}$ and $y_{j,t,[p+1,p']}$ be the methylation states $y \in \{u, p, m\}$ in non-island $j \in \{1, \cdots ,J\}$ at tip $t$ for the given segments. 

We define the mean correlation as:

\[
\overline{Cor(x)_{s,l}}=\frac1I\frac1T\sum_{i=1}^{I}\sum_{t=1}^{T}Cor(x_{i,t,[p:p'-1]}, x_{i,t,[p+1:p']}) \\
\overline{Cor(y)_{s,l}}=\frac1J\frac1T\sum_{j=1}^{J}\sum_{t=1}^{T}Cor(y_{j,t,[p:p'-1]}, y_{j,t,[p+1:p']})
\]

where correlations are computed only for segment pairs where the standard deviation of each segment is non-zero:

\[
\sigma(x_{i,t,[p:p'-1]}) \neq 0 \text{ and } \sigma(x_{i,t,[p+1:p']})\neq 0 \\
\sigma(y_{j,t,[p:p'-1]}) \neq 0 \text{ and } \sigma(y_{j,t,[p+1:p']})\neq 0
\]

In this formulation:

- Only structures with a minimum length of $n = s + 2 \cdot l$ are considered.

- If $l = 0$, all the positions are considered.

These values can be computed with the functions `compute_meanCor_i` for island structures and `compute_meanCor_ni` for non-island structures as in the following examples:

```{r}
# 1 tip / sample / replicate
sample_n <- 1
index_islands <- c(1, 3)
index_nonislands <- c(2, 4) 
data <- list(c(.5, 0, 0, 0, .5, .5, .5, .5, .5, 1, .5, 0, 0, 0, .5, .5, .5, .5, .5, 1, .5, 0, 0, 0, .5, .5, .5, .5, .5, 1), # 30 sites
               c(.5, 1, 1, 1, .5, .5, 1, 1, 1, .5, .5, 1, 1, 1, .5, .5, 1, 1, 1, .5, .5, 1, 1, 1, .5), # 25 sites
               c(.5, 0, 0, .5, 1, .5, 0, 0, .5, 1, .5, 0, 0, .5, 1, .5, 0, 0, .5, 1, .5, 0, 0, .5, 1, .5, 0, 0, .5, 1, .5, 0, 0, .5, 1, .5, 0, 0, .5, 1), # 40 sites
               c(1, 1, 1, .5, .5, .5, 0, 0, 0, .5, 1, 1, 1, .5, .5, .5, 0, 0, 0, .5, 1, 1, 1, .5, .5, .5, 0, 0, 0, .5, .5, 0, 0, 0, .5)) # 35 sites
compute_meanCor_i(index_islands, minN_CpG = 10, shore_length = 5, data, sample_n = 1)
compute_meanCor_ni(index_nonislands, minN_CpG = 10, shore_length = 5, data, sample_n = 1)
```

```{r}
# 2 tip / sample / replicate
sample_n <- 2
index_islands <- c(1, 3)
index_nonislands <- c(2, 4)
data <- list(
  # tip 1
    list(c(.5, 0, 0, 0, .5, .5, .5, .5, .5, 1, .5, 0, 0, 0, .5, .5, .5, .5, .5, 1, .5, 0, 0, 0, .5, .5, .5, .5, .5, 1), # 30 sites
         c(.5, 1, 1, 1, .5, .5, 1, 1, 1, .5, .5, 1, 1, 1, .5, .5, 1, 1, 1, .5, .5, 1, 1, 1, .5), # 25 sites
         c(.5, 0, 0, .5, 1, .5, 0, 0, .5, 1, .5, 0, 0, .5, 1, .5, 0, 0, .5, 1, .5, 0, 0, .5, 1, .5, 0, 0, .5, 1, .5, 0, 0, .5, 1, .5, 0, 0, .5, 1), # 40 sites
         c(1, 1, 1, .5, .5, .5, 0, 0, 0, .5, 1, 1, 1, .5, .5, .5, 0, 0, 0, .5, 1, 1, 1, .5, .5, .5, 0, 0, 0, .5, .5, 0, 0, 0, .5)), # 35 sites
  # tip 2
    list(c(.5, 0, 0, .5, .5, .5, 0, 0, .5, 1, .5, 0, 0, 0, 0, .5, .5, 1, 1, 1, .5, 0, 0, 0, .5, .5, 1, 1, 1, 1), # 30 sites
         c(.5, .5, 1, 1, .5, .5, 1, 1, 1, .5, .5, 0, 0, 0, .5, .5, 1, 1, 1, .5, .5, 1, 1, 1, .5), # 25 sites
         c(.5, 0, 0, .5, 1, .5, 0, 0, .5, 1, .5, 0, 0, .5, .5, .5, 0, 0, .5, 1, 1, 1, 1, .5, 1, .5, 0, 0, .5, 1, .5, 0, 0, .5, 1, .5, 0, 0, .5, 1), # 40 sites
         c(1, 1, 1, .5, .5, .5, 0, 0, 0, .5, 1, 1, 1, 1, .5, .5, 0, 0, 0, .5, 1, 1, 1, .5, .5, .5, .5, .5, 0, .5, .5, .5, .5, 0, .5)) # 35 sites
  )
compute_meanCor_i(index_islands, minN_CpG = 10, shore_length = 5, data, sample_n = 2)
compute_meanCor_ni(index_nonislands, minN_CpG = 10, shore_length = 5, data, sample_n = 2)
```


*Correlation measures how much the state of each site says about the next. So, when there is no variation (e.g. in middle segment state u of a site is followed by state u of the next), there is no correlation.*

## Comparison of tree tips

### Frequency of Methylation State Changes at Single Sites

We compare the methylation state of single sites at the tree cherries. A cherry is a pair of leaf nodes (also called tips or terminal nodes) in a phylogenetic tree that share a direct common ancestor. In other words, if two leaves are connected to the same internal node and no other leaves are connected to that internal node, they form a cherry.

We calculate the frequency of sites that changes for each structure (island and non-island structures).

Let $S_k$ be the set of sites in the cherry structure of interest, where $k$ represents the index of the cherry structure. For each site in $S_k$, let the methylation state at a given tip $t$ be denoted as $m_{k,s,t}$, where $s \in S_k$ is the index of the site, and $m_{k,s,t} \in {0, 0.5, 1}$ represents the methylation state at tip $t$ (unmethylated, partially-methylated, or methylated, respectively).

For a given cherry structure, we compute in the proportion of sites that have a different methylation state between two tips, say $t_1$ and $t_2$.

Define the indicator variable:
\[
\delta_{k,s,t_1,t_2} =
\begin{cases}
1, & \text{if } m_{k,s,t_1} \neq m_{k,s,t_2}, \\
0, & \text{if } m_{k,s,t_1} = m_{k,s,t_2}.
\end{cases}
\]

Then, the summary statistic for the proportion of sites with different methylation states between tips $t_1$ and $t_2$ in the cherry structure $k$ is:
\[
\text{Proportion}_k(t_1, t_2) = \frac{1}{|S_k|} \sum_{s \in S_k} \delta_{k,s,t_1,t_2}
\]

where $|S_k|$ is the total number of sites in the cherry structure $k$, and the sum counts the number of sites with different methylation states between tips $t_1$ and $t_2$.

This statistic gives the proportion of sites in the cherry structure $k$ where the methylation state differs between the two tips of interest.


Functions:

- `get_cherryDist(tree)` to get the distances between the tips of each cherry. It identifies each cherry by the tip names and the tip indices. The tip indices correspond to (a) the index from left to right on the newick string, (b) the order of the tip label in the `phylo_object$tip.label`, and (c) the index in the methylation data list `data[[tip]][[structure]]` as obtained with the function `simulate_evolData()` when the given tree has several tips.

- `countSites_cherryMethDiff(cherryDist, data)` to get for each cherry the counts of sites with half-methylation and full-methylation changes per genomic structure (island or non-island).

- `freqSites_cherryMethDiff(tree,data)`. The function first validates the tree structure and extracts pairwise distances between cherry tips with `get_cherryDist`. It then validates the input data and counts half and full methylation differences for each cherry at each structure using `countSites_cherryMethDiff`. Finally, it normalizes these (half and full) counts by the number of sites per structure to compute frequencies. .

- `get_siteFChange_cherry(tree,data)`. Uses `freqSites_cherryMethDiff(tree,data)` and then computes the frequency of sites with any change of methylation state (full or half) for each cherry at each genomic structure.

- `MeanSiteFChange_cherry(tree, data, index_islands, index_nonislands)`. The function computes the per-cherry frequency of sites with different methylation states at each structure (islands and non-islands) using \code{get_siteFChange_cherry}. 
Then it calculates the weighted mean site frequency of methylation changes for each cherry separately for islands and non-islands.

```{r}
# Set example tree and methylation data
tree <- "((a:1.5,b:1.5):2,(c:2,d:2):1.5);"
data <- list(
    list(rep(1,10), rep(0,5), rep(1,8)), # tip a
    list(rep(1,10), rep(0.5,5), rep(0,8)), # tip b
    list(rep(1,10), rep(0.5,5), rep(0,8)), # tip c
    list(c(rep(0,5), rep(0.5, 5)), c(0, 0, 1, 1, 1), c(0.5, 1, rep(0, 6)))) # tip d 
# Set the index for islands and non-island structures
index_islands <- c(2)
index_nonislands <- c(1, 3)

MeanSiteFChange_cherry(data = data, 
                       tree = tree, 
                       index_islands = index_islands, 
                       index_nonislands = index_nonislands)
```



